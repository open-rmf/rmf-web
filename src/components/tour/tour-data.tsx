import { ReactourStep } from 'reactour';

export const tourText: { [stepName: string]: { id: number; text: string } } = {
  Welcome: { id: 1, text: 'Welcome to RoMi dashboard' },
  Zoom: {
    id: 2,
    text:
      'Click on the zoom buttons to change the view of the floor plan. Alternatively, the scroll button on your mouse would work too!',
  },
  Floorplan: {
    id: 3,
    text:
      'Use the floor plan button to switch between available levels and enabling / disabling the view of different components.',
  },
  Leaflet: {
    id: 4,
    text:
      'Clicking individual components like doors, robots, lifts on the map will open up its corresponding information tab in the Omnipanel.',
  },
  Omnipanel: {
    id: 5,
    text: 'The Omnipanel button on the Dashboard shows the various panel options available.',
  },
  MainMenu: {
    id: 6,
    text:
      'Each Panel is clickable and contains a list of the available items with their corresponding states.',
  },
  DoorPanel: { id: 7, text: 'Let us take a look into the ðŸšª Doors Panel.' },
  DoorTab: { id: 8, text: 'Here is an example of what you will see when a door tab is expanded' },
  CommandsPanel: {
    id: 9,
    text:
      'The Commands Panel allows you to send different types of requests that will be handled by RoMi.',
  },
  LoopRequest: {
    id: 10,
    text:
      'An example is the Loop Request, where RoMi will assign the most suitable robot to perform the task at the point of request.',
  },
  Settings: {
    id: 11,
    text: 'The Settings button opens up the drawer for different dashboard settings.',
  },
  TrajAnims: {
    id: 12,
    text: 'Trajectory configurations can be changed using the options available',
  },
  HelpButton: {
    id: 13,
    text: 'The Help button opens up another drawer with useful tools and resources',
  },
  HelpDrawer: {
    id: 14,
    text: `Check out the Hotkey options available and the Tutorial tab opens this guide again. Look out for new features ahead!`,
  },
};

export type stepStyling = {
  backgroundColor: string;
  color: string;
  borderRadius: string;
};

export type stepDetails = {
  [primaryKey: string]: ReactourStep;
};
