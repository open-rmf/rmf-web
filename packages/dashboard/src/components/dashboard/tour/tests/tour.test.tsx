import { createMount, createShallow } from '@material-ui/core/test-utils';
import { ReactWrapper } from 'enzyme';
import React from 'react';
import DashboardTour from '../tour';
import { tourText } from '../tour-data';

/*react-leaflet does not work well on jsdom,
suppress warnings of not being able to find elements from leaflet*/
jest.spyOn(console, 'warn').mockImplementation(() => {});

const mount = createMount();
const shallow = createShallow();

describe('Dashboard Tour', () => {
  const tourState: boolean = true;
  const setTourState: jest.Mock = jest.fn();
  const setShowSettings: jest.Mock = jest.fn();
  const setShowOmniPanel: jest.Mock = jest.fn();
  const setShowHelp: jest.Mock = jest.fn();
  const clearSpotlights: jest.Mock = jest.fn();
  const setCurrentView: jest.Mock = jest.fn();
  const doorSpotlight: jest.Mock = jest.fn();

  let root: ReactWrapper;
  let expectedText: string;
  let stepText: string;
  let nextBtn: ReactWrapper;
  const tourProps = {
    tourState,
    setTourState,
    setShowSettings,
    setShowOmniPanel,
    setShowHelp,
    clearSpotlights,
    setCurrentView,
    doorSpotlight,
  };
  const wait = (delay = 0) => new Promise((resolve) => setTimeout(resolve, delay));

  beforeEach(() => {
    root = mount(<DashboardTour tourProps={tourProps} />);
    nextBtn = root.find('[data-testid="next-btn"]').first();
  });

  afterEach(() => {
    root?.unmount();
  });

  it('renders without crashing', () => {
    const tour = shallow(<DashboardTour tourProps={tourProps} />);
    expect(tour).toMatchSnapshot();
  });

  it('renders first welcome step', () => {
    expectedText = tourText.Welcome.text;
    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders zoom button step', async () => {
    expectedText = tourText.Zoom.text;

    nextBtn.simulate('click');
    await wait(5);

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders floor plan button step', async () => {
    expectedText = tourText.Floorplan.text;

    for (let i = 1; i < tourText.Floorplan.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders leaflet step', async () => {
    expectedText = tourText.Leaflet.text;

    for (let i = 1; i < tourText.Leaflet.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders omnipanel button step', async () => {
    expectedText = tourText.Omnipanel.text;

    for (let i = 1; i < tourText.Omnipanel.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders main menu step', async () => {
    expectedText = tourText.MainMenu.text;

    for (let i = 1; i < tourText.MainMenu.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders door panel step', async () => {
    expectedText = tourText.DoorPanel.text;

    for (let i = 1; i < tourText.DoorPanel.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders door tab step', async () => {
    expectedText = tourText.DoorTab.text;

    for (let i = 1; i < tourText.DoorTab.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders commands panel step', async () => {
    expectedText = tourText.CommandsPanel.text;

    for (let i = 1; i < tourText.CommandsPanel.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders loop request step', async () => {
    expectedText = tourText.LoopRequest.text;

    for (let i = 1; i < tourText.LoopRequest.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders settings button step', async () => {
    expectedText = tourText.Settings.text;

    for (let i = 1; i < tourText.Settings.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders traj anims step', async () => {
    expectedText = tourText.TrajAnims.text;

    for (let i = 1; i < tourText.TrajAnims.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders help button step', async () => {
    expectedText = tourText.HelpButton.text;

    for (let i = 1; i < tourText.HelpButton.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('renders help drawer step', async () => {
    expectedText = tourText.HelpDrawer.text;

    for (let i = 1; i < tourText.HelpDrawer.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    stepText = root.find('[data-testid="stepBox"]').first().text();

    expect(stepText.includes(expectedText));
  });

  it('closes the tour when last step button is clicked', async () => {
    for (let i = 0; i < tourText.HelpDrawer.id; i++) {
      nextBtn.simulate('click');
      await wait(5);
    }

    root.find('[data-testid="last-step-btn"]').first().simulate('click');

    expect(setTourState).toHaveBeenCalledTimes(1);
  });
});
