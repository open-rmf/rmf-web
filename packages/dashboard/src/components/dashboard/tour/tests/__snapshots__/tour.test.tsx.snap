// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Tour renders without crashing 1`] = `
<Tour
  accentColor="#3f51b5"
  badgeContent={[Function]}
  closeWithMask={false}
  disableFocusLock={false}
  disableInteraction={false}
  disableKeyboardNavigation={
    Array [
      "right",
      "left",
    ]
  }
  isOpen={true}
  maskSpace={10}
  onRequestClose={[Function]}
  rounded={5}
  scrollDuration={1}
  showButtons={false}
  showCloseButton={true}
  showNavigation={false}
  showNavigationNumber={false}
  showNumber={true}
  startAt={0}
  steps={
    Array [
      Object {
        "action": [Function],
        "content": [Function],
        "selector": "",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": [Function],
        "content": [Function],
        "selector": "[class=\\"leaflet-control-zoom leaflet-bar leaflet-control\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": [Function],
        "content": [Function],
        "selector": "[class= \\"leaflet-control-layers leaflet-control\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": [Function],
        "content": [Function],
        "selector": "[class=\\"leaflet-image-layer leaflet-zoom-animated\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[id=\\"toggle-omnipanel-btn\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[data-component=\\"MainMenu\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[data-item=\\"Doors\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[data-name=\\"main_door\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[data-component=\\"LoopForm\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": ".MuiDrawer-paper",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[id=\\"show-help-btn\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
      Object {
        "action": undefined,
        "content": [Function],
        "selector": "[id=\\"help-drawer-options\\"]",
        "style": Object {
          "backgroundColor": "#3f51b5",
          "borderRadius": "5px",
          "color": "#fff",
        },
      },
    ]
  }
  update="0"
  updateDelay={1}
/>
`;
